main.c

```diff
@@ -252,6 +294,7 @@ MPZ_from_str(PyObject *s, int base)
     Py_ssize_t len;
     const char *str = PyUnicode_AsUTF8AndSize(s, &len);
 
+    underscores_as_number((char *)str, &len);
```

Comment:
You should check return value.

Also, please read Python docs about PyUnicode_* functions.  You did this in a wrong way, for example:
```pycon
>>> from gmp import mpz
>>> s = '1_2_3'
>>> mpz(s)
mpz(123)
>>> s
'123\x003'
```

Reply:
Hmmm, it seems that I missed something. I'll check them.

Reply:
MPZ_from_str() code below already allocates a temporary buffer to handle prefixes, etc.  I would rather land such changes there.

Please also take look on the CPython code.

Reply:
I will carefully study the unicode implementation in CPython.